lib/Sema/SemaBounds.cpp

```diff
@@ -1835,19 +1935,24 @@ namespace {
       }
       exit:
         return I;
-     }
+    }
 
-    // Convert an expression E into a base and offset form.  The offset is
-    // in characters.
+    // Convert an expression E into a base and offset form.
     // - If E is pointer arithmetic involving addition or subtraction of a
     //   constant integer, return the base and offset.
+    // - If E is a pointer arithmetic involving addition or subtraction of
+    //   a pointer with an unsigned integer expression, return the pointer as
+    //   the base, and the rest as the offset. (If OnlyConstant is set to true,
+    //   this case is not checked. Instead, the function will try to split the
+    //   E into Base and Offset assuming the Offset can only be constant.)
     // - If it is not or we run into issues such as pointer arithmetic overflow,
     // return a default expression of (E, 0).
     // TODO: we use signed integers to represent the result of the Offset.
     // We can't represent unsigned offsets larger the the maximum signed
     // integer that will fit pointer width.
     void SplitIntoBaseAndOffset(Expr *E, Expr *&Base,
-                                llvm::APSInt &Offset) {
+                                llvm::APSInt &Offset, Expr *&OffsetExpr,
```

Comment:
Instead of adding the extra parameter OnlyConstant, I suggest always letting this function always split expressions if possible and checking the results returned by the function.   I think you will need to change the function to return a value indicating what it did.